rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,212 @@
+- Start Date: 2020-06-02
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+Pluggable state from registry items can often end up stored inside saved objects, for instance:
+Embeddables, Expression strings, Url generators, ...
+
+When plugin A (persister) stores some state that belongs to another plugin B a few issues arise:
+- How does persister know if plugin B state contains references to saved objects
+- How does persister migrate the saved object when it contains state that belongs to plugin B
+- How does persister know if state that belongs to plugin B contains state that belongs to plugin C
+
+# Basic example
+
+```ts
+
+const MY_STATE_ID = 'MyState';
+
+interface MyStateV77 extends PersistableState {
+  objectId: string,
+  val: number,
+}
+
+interface MyState extends PersistableState {
+  objectId: string,
+  value: number,
+}
+
+const migrate = (state: MyState, version: number) => {
+  let newState: MyState,
+  if (version < 7.8) {
+    const oldState = state as MyStateV77;
+    newState = { objectId: oldState.objectId, value: oldState.val };
+  } else {
+    newState = state;
+  }
+  
+  return newState;
+}
+
+const inject = (state: MyState, savedObjectReferences: SavedObjectReference[]) => {
+  return {
+    ...state,
+    objectId: savedObjectReferences.find(ref => ref.name = state.objectId)?.id;
+  }
+}
+
+const extract = (state: MyState) => {
+  const references = [{ name: 'objectId', id: state.objectId, type: 'savedObject' }];
+  return [{ ...state, objectId: 'objectId' }, references];
+}
+
+persistableStateService.register(MY_STATE_ID, { migrate, inject, extract });
+```
+
+```ts
+  const stateReadyForPersisting = persistableStateService.get(MY_STATE_ID),beforeSave(myState);
+  const stateReadyForConsuming = persistableStateService.get(MY_STATE_ID).afterLoad(stateReadyForPersisting);
+```
+
+# Motivation
+
+We need a formal way for authors of registry implementations to add migrations for this data. Since they may not own the saved object that their implementation ends up in, we can't rely on saved object migrations as it exists today (not to mention data that ends up in URLs that may need to be migrated).
+
+We also need to make sure that all the persited state containing references to saved objects extracts those before being saved and injects them later.
+
+# Detailed design
+
+```ts
+interface PersistableState extends Serializable {}
+
+interface PersistableStateDefinition {
```

Comment:
as we will be recommending to import the right utilities directly from plugins when you know what state you are consuming this should not be needed. If this should only be used when you don't know the state id at compile time, then in those cases you will always work with PersistableState type as you won't have any other information about the state you are working with.

if we decide to recommend to use the service even when you know id at compile time but rather just import the types from correct plugin then this would make sense. but why would we do that ? if you are already importing form the plugin you might as well import the migration utility ?

if you want to use the service due to easier handling of state (beforeLoad and afterLoad methods), you can still use them and cast.

Reply:
I agree with importing utilities direct from plugins when possible.  It still creates an awkward experience for a plugin author when registering their function.

How could an implementor register their PersistableStateMigrator without casting the passed in state? `PersistableState` is typed as a serializable empty object:

```ts
const inject = (state: {}, references: SavedObjectReference[]): {} => {
  const myRealState = state as SomeThingElse;  
  // do injection.

  // In what situation is this return value typed as {} useful?
  return myRealState as {};
} 
```

Are you saying every implementor will have to internally cast, in each function, from `{}` to the real `PersistableState` shape?

Reply:
Your example below also won't work with this typescript.  Example:

```ts
interface MyStateLatest {
  foo: string
}

interface MyStateV1 {
  bar: string
}

const inject = (state: MyStateLatest, savedObjectReferences: any[]) => {
  // No typescript error, but shouldn't be allowed.  State should match `MyStateLatest`
  return {};
}

type PersistableState = { }

interface PersistableStateDefinition {
  id: string,
  // migrate function receives state and version string and should return latest state version
  // default is identity function
  //migrate: (state: PersistableState, version: string) => PersistableState,
  // inject function receives state and a list of references and should return state with references injected
  // default is identity function
  inject: (state: PersistableState, references: any[]) => PersistableState,
  // extract function receives state and should return state with references extracted and array of references
  // default returns same state with empty reference array
  //extract: (state: PersistableState) => { state: PersistableState, references: any[] }
}

const myDef: PersistableStateDefinition = {
  //  Property 'foo' is missing in type 'PersistableState' but required in type 'MyStateLatest'.(2322)
  inject,
}




```

[typescript playground](
https://www.typescriptlang.org/play/?ssl=29&ssc=103&pln=28&pc=44#code/JYOwLgpgTgZghgYwgAgLIE8DKY6QDK4QDOYyA3gLABQyyMA9vQFzIlSgDm1AvtdaJFiIUGbIQBqARnLVaAIzhQWbTjz5UE9ECWSgAVhASkAvMgAUJQi1E58hEgBpWcAG4QAJgHk5BowCUIGGgIECQiFjgQdABtAF0ASmRjAD4ZGmQoCDAAVygQcm4AbjUqajB0AAcUAAVoImBLOQAbCDFIJPJkXlKqAWh4JGRaqHrGlraIABFA0GAwYC002mB3ZTB2EA4HWWQAel3kAFtgDihCOmzQ+cXMpGA3IlZbFEj3ZDcRhfyVTeRX1gAFvRsk03pkcnlkE17KRLO0PvUtDt9sh3IE4CDSA1dGjwHN0Bcrl9kbtjqcrOY4RAWMNRjhmq1nk4EV81hsOIkUkM6g16eMmSTdCBfKQYJcjF8MoYIPdiE9zv84FDech6DApUFMqE5f8iECQWCsrlvs9kAB3OYAjXBbWPfSGSDuQVo+CY3R23HzcqEiVI9L2owsCzPGk8saMwhOTKakJhCJROKc1K03lwBkTbbpFEQAAe60QovF13ytxlD3l7V1+tBUohJvOFrAVujNrCyFz+aMHj+IDeijOBLV1q1YWd6Ld4ONjyIcEOKCp5st7cOFW9LZHLygA5JHbOgcpIe5n3DEyTnSpoePfIjkCjgVbxHjMViXRK1E02lIh3Q0xgl7pab8oQv6zMWHSUFmBxDFA9BVFA3oAOQMPQCHukcDT1L8oDIOUVTIAhKYns8qFyNkpCZAAjtkwCZG82G4SgCE2IQBCQCQCEAHRmAATAAzNx3HxDsAZgJm3TqNQQA)

Reply:
as the implementator i will do something like this :
```
persistableState.register('vis', {
  inject: (state: SerializableVis, references: SavedObjectReference[]) => {
     // no casting needed, state is SerializableVis
   }
```

all i need to make sure is that SerializableVis implements PersistableState (which means it needs to be serializable)

Reply:
Can you verify this on typescript playground? It doesn't work for me, I get the error 

```
 //  Property 'foo' is missing in type 'PersistableState' but required in type 'MyStateLatest'.(2322)
```

See the link above.